sklearn/preprocessing/_discretization.py

```diff
@@ -240,19 +240,28 @@ def fit(self, X, y=None, sample_weight=None):
 
         if self.subsample is not None and n_samples > self.subsample:
             # Take a subsample of `X`
+            # When resampling, it important to subsample **with replacement** to
+            # preserve the distribution, in particular in the presence of a few data
+            # points with large weights.
             X = resample(
                 X,
-                replace=False,
+                replace=True,
                 n_samples=self.subsample,
                 random_state=self.random_state,
+                sample_weight=sample_weight,
             )
+            # Since we already used the weights when resampling when provided,
+            # we set them back to `None` to avoid accounting for the weights twice
+            # in subsequent operations to compute weight-aware bin edges with
+            # quantiles or k-means.
+            sample_weight = None
+
+        elif sample_weight is not None:
+            sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
```

Comment:
it's a different matter
```suggestion
        if sample_weight is not None:
            sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
```

Reply:
Indeed, we still need to call `_check_sample_weight` when `KBinsDiscretizer.fit` is called with a valid non-numpy input such as a Python list `sample_weight=[1, 2, 3]` and subsampling is disabled for instance.